gdal/frmts/netcdf/netcdfsg.h

```diff
@@ -0,0 +1,299 @@
+	};
+	
+	// Failed Read
+	class SG_Exception_BadPoint : public SG_Exception
+	{
+		const char * err_msg;
+
+		public:
+			char* get_err_msg() override { return (char*)err_msg; }
+		
+		SG_Exception_BadPoint() : err_msg("An attempt was made to read an invalid point (likely index out of bounds).") {}
+	};
+
+	// Too many dimensions on node coordinates variable
+	class SG_Exception_Not1D : public SG_Exception
+	{
+		const char * err_msg;
+
+		public:
+			char* get_err_msg() override { return (char*)err_msg; }
+		
+		SG_Exception_Not1D() : err_msg("A node coordinates axis variable is not one dimensional.") {}
+	};
+
+	// Some helpers which simply call some netcdf library functions, unless otherwise mentioned, ncid, refers to its use in netcdf.h
+	
+	/* Retrieves the minor version from the value Conventions global attr
+	 * Returns: a positive integer corresponding to the conventions value
+	 *	if not CF-1.x then return negative value, -1
+	 */
+	int getCFMinorVersion(int ncid);
+
+	/* Given a geometry_container varID, searches that variable for a geometry_type attribute
+	 * Returns: the equivalent geometry type
+	 */
+	geom_t getGeometryType(int ncid, int varid); 
+	
+	void* inPlaceSerialize_Point(SGeometry * ge, int seek_pos, void * serializeBegin);
+	void* inPlaceSerialize_LineString(SGeometry * ge, int node_count, int seek_begin, void * serializeBegin);
+	void* inPlaceSerialize_PolygonExtOnly(SGeometry * ge, int node_count, int seek_begin, void * serializeBegin);
+	void* inPlaceSerialize_Polygon(SGeometry * ge, std::vector<int>& pnc, int ring_count, int seek_begin, void * serializeBegin);
+
+	/* scanForGeometryContainers
+	 * A simple function that scans a netCDF File for Geometry Containers
+	 * -
+	 * Scans the given ncid for geometry containers
+	 * The vector passed in will be overwritten with a vector of scan results
+	 */
+	int scanForGeometryContainers(int ncid, std::vector<int> & r_ids);
+
+	/* Given a variable name, and the ncid, returns a SGeometry reference object, which acts more of an iterator of a geometry container
+	 *
+	 * Returns: a NEW geometry "reference" object, that can be used to quickly access information about the object, nullptr if invalid or an error occurs
+	 */
+	SGeometry* getGeometryRef(int ncid, const char * varName );	
+
+	/* Writes a geometry to the specified NC file.
+	 * The write uses the following guidelines
+	 * - the geometry variable name is geometryX where X is some natural number
+	 * - the geometry container is called geometry_containerX, corresponding to some geometryX
+	 *
+	 * Returns: an error code if failure, 0 on success
+	 */
+	//int putGeometryRef(int ncid, SGeometry* geometry);
+
+	/* Fetches a one dimensional string attribute
+	 * using the given ncid, variable ID (varID), and attribute key (attrName)
+	 *
+	 * Returns: a NEW cstring of the attribute value, that must be separately deleted 
+	 */
+	char * attrf(int ncid, int varId, const char * attrName);
```

Comment:
would be more simple to return a std::string to avoid dealing with memory deallocation

Reply:
like the C++-styled recommendations you've given! :+1: okay will fix